File: nodejs/patterns/src/behavioral/memento.js
class Memento {
  constructor(text) {
    this.text = text;
  }
}

class TextEditor {
  constructor() {
    this.text = "";
  }

  type(value) {
    this.text = value;
  }

  save() {
    return new Memento(this.text);
  }

  restore(memento) {
    this.text = memento.text;
  }
}

function demo() {
  const editor = new TextEditor();
  editor.type("draft");
  const checkpoint = editor.save();
  editor.type("final");
  editor.restore(checkpoint);
  return editor.text;
}

module.exports = { Memento, TextEditor, demo };

File: python/patterns/src/prototype_patterns/behavioral/memento.py
"""Memento pattern example."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Memento:
    text: str


class TextEditor:
    def __init__(self) -> None:
        self._text = ""

    def type(self, value: str) -> None:
        self._text = value

    def save(self) -> Memento:
        return Memento(self._text)

    def restore(self, memento: Memento) -> None:
        self._text = memento.text

    @property
    def text(self) -> str:
        return self._text


def demo() -> str:
    editor = TextEditor()
    editor.type("draft")
    checkpoint = editor.save()
    editor.type("final")
    editor.restore(checkpoint)
    return editor.text
